okio-wasifilesystem/src/wasmMain/kotlin/okio/internal/preview1/Wasi.kt

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2023 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@file:OptIn(UnsafeWasmMemoryApi::class)
+
+package okio.internal.preview1
+
+import kotlin.wasm.unsafe.MemoryAllocator
+import kotlin.wasm.unsafe.Pointer
+import kotlin.wasm.unsafe.UnsafeWasmMemoryApi
+import kotlin.wasm.unsafe.withScopedMemoryAllocator
+import okio.Path.Companion.toPath
+
+
+internal fun pathOpen(path: String, oflags: oflags, rightsBase: rights): fd {
+  withScopedMemoryAllocator { allocator ->
+    val (pathAddress, pathSize) = allocator.write(path)
+
+    val returnPointer: Pointer = allocator.allocate(4) // fd is u32.
+    val errno = path_open(
+      fd = UnspecifiedFd,
+      dirflags = 0,
+      path = pathAddress.address.toInt(),
+      pathSize = pathSize,
+      oflags = oflags,
+      fs_rights_base = rightsBase,
+      fs_rights_inheriting = 0,
+      fdflags = 0,
+      returnPointer = returnPointer.address.toInt(),
+    )
+    if (errno != 0) throw ErrnoException(errno.toShort())
+    return returnPointer.loadInt()
+  }
+}
+
+@OptIn(UnsafeWasmMemoryApi::class)
+internal fun fdReadDir(fd: fd): List<DirectoryEntry> {
+  withScopedMemoryAllocator { allocator ->
+    var bufSize = 2048
+    var bufPointer = allocator.allocate(bufSize)
+    val returnPointer = allocator.allocate(4) // `size` is u32, 4 bytes.
+    var pageSize: Int
+
+    // In theory, fd_readdir uses a 'cookie' field to page through results. In practice the NodeJS
+    // implementation doesn't honor the cookie and directories with large file names don't progress.
+    // Instead, just grow the buffer until the entire directory fits.
```

Comment:
FYI @sdeleuze not sure if you ran into similar problems. I tested this on some long file names and it got very grumpy very fast.